x/zkism/types/metadata.go

```diff
@@ -97,3 +107,122 @@ func NewZkExecutionISMMetadata(metadata []byte) (ZkExecutionISMMetadata, error)
 func (meta ZkExecutionISMMetadata) HasExecutionProof() bool {
 	return len(meta.Proof) > 0
 }
+
+// PublicInputs defines the set of public values associated with a proof generated by a ZK program.
+type PublicInputs struct {
+	// TrustedStateRoot is the trusted state root from which we are updating.
+	TrustedStateRoot [32]byte
+	// NewHeaderHash is the block header hash of the latest state update.
+	NewHeaderHash [32]byte
+	// PreviousHeaderHash is the block header hash of the previous state.
+	PreviousHeaderHash [32]byte
+	// CelestiaHeaderHashes is the ordered set of Celestia header hashes whose
+	// data roots contain the blobs posted between the previous and new headers.
+	// TODO: A single hash (latest celestia header hash) should be possible to use here with further changes to proof system.
+	CelestiaHeaderHashes [][]byte
+	// NewStateRoot is the resulting state root after applying the state transition.
+	NewStateRoot [32]byte
+	// NewHeight is the block height associated with the new header.
+	NewHeight uint64
+}
+
+// Marshal encodes the PublicInputs struct into a bincode-compatible byte slice.
+// The output format uses Rust bincode's default configuration: (little-endian, fixed-width integers, length-prefixed slices).
+func (pi *PublicInputs) Marshal() ([]byte, error) {
+	var buf bytes.Buffer
+
+	if err := writeBytes(&buf, pi.TrustedStateRoot[:]); err != nil {
+		return nil, fmt.Errorf("write TrustedStateRoot: %w", err)
+	}
+	if err := writeBytes(&buf, pi.NewHeaderHash[:]); err != nil {
+		return nil, fmt.Errorf("write NewHeaderHash: %w", err)
+	}
+	if err := writeBytes(&buf, pi.PreviousHeaderHash[:]); err != nil {
+		return nil, fmt.Errorf("write PreviousHeaderHash: %w", err)
+	}
+
+	if err := binary.Write(&buf, binary.LittleEndian, uint64(len(pi.CelestiaHeaderHashes))); err != nil {
+		return nil, fmt.Errorf("write CelestiaHeaderHashes length: %w", err)
+	}
+
+	for i, hash := range pi.CelestiaHeaderHashes {
+		if len(hash) != 32 {
+			return nil, fmt.Errorf("CelestiaHeaderHashes[%d] has invalid length %d (expected 32)", i, len(hash))
+		}
+		if err := writeBytes(&buf, hash); err != nil {
+			return nil, fmt.Errorf("write CelestiaHeaderHashes[%d]: %w", i, err)
+		}
+	}
+
+	if err := writeBytes(&buf, pi.NewStateRoot[:]); err != nil {
+		return nil, fmt.Errorf("write NewStateRoot: %w", err)
+	}
+
+	if err := binary.Write(&buf, binary.LittleEndian, pi.NewHeight); err != nil {
+		return nil, fmt.Errorf("write NewHeight: %w", err)
+	}
+
+	return buf.Bytes(), nil
+}
+
+// Unmarshal decodes a bincode-serialized PublicInputs struct.
+// This function expects the input byte slice to be encoded using Rust bincode's
+// default configuration: (little-endian, fixed-width integers, length-prefixed slices).
+func (pi *PublicInputs) Unmarshal(data []byte) error {
```

Comment:
The bincode marshalling and unmarshalling is a bit verbse but such is life. But afaict it needs to be done to check validity of pub inputs and to pass the correct encoding format the proof expects